File: backend/src/main/java/lan/dk/podcastserver/manager/worker/sixplay/SixPlayUpdater.java
package lan.dk.podcastserver.manager.worker.sixplay;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.github.davinkevin.podcastserver.service.HtmlService;
import com.github.davinkevin.podcastserver.service.ImageService;
import com.github.davinkevin.podcastserver.service.SignatureService;
import com.jayway.jsonpath.DocumentContext;
import com.jayway.jsonpath.TypeRef;
import io.vavr.Value;
import io.vavr.collection.HashMap;
import io.vavr.collection.HashSet;
import io.vavr.collection.Set;
import io.vavr.control.Option;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.exception.parser.SixPlayParsingException;
import lan.dk.podcastserver.manager.worker.Type;
import lan.dk.podcastserver.manager.worker.Updater;
import lan.dk.podcastserver.service.JsonService;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

import static io.vavr.API.Option;
import static io.vavr.API.Try;
import static java.util.Objects.nonNull;

/**
 * Created by kevin on 20/12/2016 for Podcast Server
 */
@Slf4j
@Component("SixPlayUpdater")
@RequiredArgsConstructor
public class SixPlayUpdater implements Updater {

    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss", Locale.ENGLISH);
    private static final TypeRef<Set<SixPlayItem>> TYPE_ITEMS = new TypeRef<Set<SixPlayItem>>(){};
    private static final TypeRef<HashMap<String, Object>> TYPE_KEYS = new TypeRef<HashMap<String, Object>>(){};

    private static final String URL_TEMPLATE_PODCAST = "http://www.6play.fr/%s-p_%d/";
    private static final String PROGRAM_CODE_SELECTOR = "program.programsById.%d.code";
    private static final String PROGRAM_ID_SELECTOR = "program.programsById";
    private static final String VIDEO_BY_ID_SELECTOR = "video.programVideoById[*]";

    private final SignatureService signatureService;
    private final HtmlService htmlService;
    private final JsonService jsonService;
    private final ImageService imageService;

    @Override
    public Set<Item> getItems(Podcast podcast) {
        return Try(() ->
                htmlService.get(podcast.getUrl())
                        .map(d -> this.extractItems(d.select("script")))
                        .getOrElse(HashSet::empty)
        ).getOrElseThrow(SixPlayParsingException::new);
    }

    private Set<Item> extractItems(Elements script) {
        Option<DocumentContext> root6Play = extractJson(script);

        Integer programId = root6Play
                .map(JsonService.to(PROGRAM_ID_SELECTOR, TYPE_KEYS))
                .map(HashMap::keySet)
                .flatMap(Value::toOption)
                .map(Integer::valueOf)
                .getOrElseThrow(() -> new RuntimeException("programId not found in react store"));

        String programCode = root6Play
                .map(JsonService.to(String.format(PROGRAM_CODE_SELECTOR, programId), String.class))
                .getOrElseThrow(() -> new RuntimeException("programCode not found in react store"));

        String basePath = String.format(URL_TEMPLATE_PODCAST, programCode, programId);

        return root6Play
                .map(JsonService.to(VIDEO_BY_ID_SELECTOR, TYPE_ITEMS))
                .getOrElse(HashSet::empty)
                .map(s -> this.convertToItem(s, basePath));
    }

    private Item convertToItem(SixPlayItem i, String basePath) {
        return Item.builder()
                .title(i.getTitle())
                .pubDate(i.getLastDiffusion())
                .length(i.getDuration())
                .url(i.url(basePath))
                .description(i.getDescription())
                .cover(imageService.getCoverFromURL(i.cover()))
                .build();
    }

    private Option<DocumentContext> extractJson(Elements elements) {
        return getRoot6Play(elements)
                .map(this::removeJS)
                .map(jsonService::parse);
    }

    private String removeJS(String s) {
        return Option(s)
                .map(String::trim)
                .map(c -> c.replaceAll("function [^}]*", "{}"))
                .map(c -> StringUtils.removeEnd(c, ";"))
                .getOrElse(s);
    }

    public static Option<String> getRoot6Play(Elements elements) {
        return HashSet.ofAll(elements)
                .find(s -> s.html().contains("root."))
                .map(Element::html)
                .map(s -> StringUtils.substringBetween(s, " = ", "}(this));"));
    }

    @Override
    public String signatureOf(Podcast podcast) {
        return Try(() ->
                htmlService.get(podcast.getUrl())
                        .map(d -> d.select("script"))
                        .flatMap(this::extractJson)
                        .map(JsonService.extract("video.programVideosBySubCategory"))
                        .map(Object::toString)
                        .map(signatureService::fromText)
                        .getOrElseThrow(() -> new RuntimeException("Error during signature of podcast " + podcast.getTitle()))
        ).getOrElseThrow(SixPlayParsingException::new);
    }

    @Override
    public Type type() {
        return new Type("SixPlay", "6Play");
    }

    @Override
    public Integer compatibility(String url) {
        return isFrom6Play(url);
    }

    public static int isFrom6Play(String url) {
        return nonNull(url) && url.contains("www.6play.fr/") ? 1 : Integer.MAX_VALUE;
    }

    @EqualsAndHashCode
    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class SixPlayItem{

        @Getter @Setter Image display_image;
        @Getter @Setter String code;
        @Getter @Setter String description;
        @Getter @Setter String title;
        @Setter String lastDiffusion; /* 2016-12-18 11:20:00 */
        @Getter @Setter Long duration;
        @Getter @Setter String id;

        ZonedDateTime getLastDiffusion() {
            return Option(lastDiffusion)
                    .map(l -> ZonedDateTime.of(LocalDateTime.parse(l, DATE_FORMATTER), ZoneId.of("Europe/Paris")))
                    .getOrElse(ZonedDateTime::now);
        }

        String url(String basePath) {
            return basePath + code + this.shortId();
        }

        String cover() {
            return Option(display_image)
                    .map(Image::url)
                    .getOrElse(() -> null);
        }

        private String shortId() {
            return Option(id)
                    .map(s -> s.substring(0, 1))
                    .map(v -> "-" + v + "_")
                    .map(v -> v + StringUtils.substringAfter(id, "_"))
                    .getOrElse("");
        }

        @JsonIgnoreProperties(ignoreUnknown = true)
        private static class Image {

            private static final String domain = "https://images.6play.fr";
            private static final String path = "/v1/images/%s/raw?width=600&height=336&fit=max&quality=60&format=jpeg&interlace=1";
            private static final String salt = "54b55408a530954b553ff79e98";

            @Getter @Setter Integer external_key;

            public String url() {
                String path = String.format(Image.path, external_key);
                return domain + path + "&hash=" + DigestUtils.sha1Hex(path + salt);
            }
        }
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/sixplay/SixPlayUpdater.kt
package com.github.davinkevin.podcastserver.manager.worker.sixplay

import arrow.core.Option
import arrow.core.getOrElse
import arrow.syntax.collections.firstOption
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.github.davinkevin.podcastserver.manager.worker.*
import com.github.davinkevin.podcastserver.service.HtmlService
import com.github.davinkevin.podcastserver.service.ImageService
import com.github.davinkevin.podcastserver.service.SignatureService
import com.github.davinkevin.podcastserver.utils.k
import com.jayway.jsonpath.DocumentContext
import com.jayway.jsonpath.TypeRef
import lan.dk.podcastserver.service.JsonService
import org.apache.commons.codec.digest.DigestUtils
import org.apache.commons.lang3.StringUtils
import org.jsoup.select.Elements
import org.springframework.stereotype.Component
import java.net.URI
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.*

/**
 * Created by kevin on 20/12/2016 for Podcast Server
 */
@Component
class SixPlayUpdater(private val signatureService: SignatureService, private val htmlService: HtmlService, private val jsonService: JsonService, private val imageService: ImageService) : Updater {

    override fun blockingFindItems(podcast: PodcastToUpdate): Set<ItemFromUpdate> =
            htmlService.get(podcast.url.toASCIIString())
                    .map { it.select("script") }
                    .map { extractItems(it) }
                    .getOrElse { setOf() }

    private fun extractItems(script: Elements): Set<ItemFromUpdate> {
        val root6Play = extractJson(script)
                .getOrElse { throw RuntimeException("No parsable JS found in the page") }

        val programId = JsonService.to(PROGRAM_ID_SELECTOR, TYPE_KEYS).apply(root6Play)
                .keys
                .first()
                .toInt()

        val programCode = JsonService.to(PROGRAM_CODE_SELECTOR.format(programId), String::class.java)
                .apply(root6Play)

        val basePath = URL_TEMPLATE_PODCAST
                .format(programCode, programId)

        return JsonService.to(VIDEO_BY_ID_SELECTOR, TYPE_ITEMS)
                .apply(root6Play)
                .filter { it.service_display.code in replay }
                .map { convertToItem(it, basePath) }
                .toSet()
    }

    private fun convertToItem(i: SixPlayItem, basePath: String) =
            ItemFromUpdate (
                title =  i.title!!,
                pubDate =  i.getLastDiffusion(),
                length =  i.duration,
                url =  URI(i.url(basePath)),
                description =  i.description,
                cover =  i.cover().map { imageService.fetchCoverInformation(it)}.orNull()?.toCoverFromUpdate()
            )

    private fun extractJson(elements: Elements): Option<DocumentContext> =
            elements
                    .toSet()
                    .firstOption { it.html().contains("root.") }
                    .map { it.html() }
                    .map { it.substringBetween(" = ", "}(this));")!!}
                    .map { it.removeJs() }
                    .map { jsonService.parse(it) }

    private fun String.removeJs() = this
            .trim { v -> v <= ' ' }
            .replace("function [^}]*".toRegex(), "{}")
            .removeSuffix(";")
            .cleanForJsonSerialization()

    override fun blockingSignatureOf(url: URI) =
            htmlService.get(url.toASCIIString()).k()
                    .map { it.select("script") }
                    .flatMap { extractJson(it) }
                    .map { JsonService.extract<Any>("video.programVideosBySubCategory").apply(it) }
                    .map { signatureService.fromText(it.toString()) }
                    .getOrElse { throw RuntimeException("Error during signature of podcast with url ${url.toASCIIString()}") }

    override fun type() = Type("SixPlay", "6Play")

    override fun compatibility(url: String?) = isFrom6Play(url)

    @JsonIgnoreProperties(ignoreUnknown = true)
    private data class SixPlayItem(
            var display_image: Image? = null,
            var code: String? = null,
            var description: String? = null,
            var title: String? = null,
            var lastDiffusion: String? = null /* 2016-12-18 11:20:00 */,
            var duration: Long? = null,
            var id: String? = null,
            val service_display: ServiceDisplay
    ){

        fun getLastDiffusion(): ZonedDateTime {
            return Option.fromNullable(lastDiffusion)
                    .map { ZonedDateTime.of(LocalDateTime.parse(it, DATE_FORMATTER), ZoneId.of("Europe/Paris")) }
                    .getOrElse { ZonedDateTime.now() }
        }

        fun url(basePath: String): String = "$basePath$code${shortId()}"

        fun cover() =
                Option.fromNullable(display_image)
                        .map { it.url() }

        private fun shortId() =
                Option.fromNullable(id)
                        .map { it.substring(0, 1) }
                        .map { "-${it}_" }
                        .map { it + StringUtils.substringAfter(id, "_") }
                        .getOrElse { "" }

        @JsonIgnoreProperties(ignoreUnknown = true)
        private class Image {

            var external_key: Int? = null

            fun url(): String {
                val path = path.format(external_key)
                return "$domain$path&hash=${DigestUtils.sha1Hex(path + salt)}"
            }

            companion object {
                private val domain = "https://images.6play.fr"
                private val path = "/v1/images/%s/raw?width=600&height=336&fit=max&quality=60&format=jpeg&interlace=1"
                private val salt = "54b55408a530954b553ff79e98"
            }
        }

        @JsonIgnoreProperties(ignoreUnknown = true)
        class ServiceDisplay(var code: String = "")
    }

    companion object {

        private val DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss", Locale.ENGLISH)
        private val TYPE_ITEMS = object : TypeRef<Set<SixPlayItem>>() {}
        private val TYPE_KEYS = object : TypeRef<Map<String, Any>>() {}

        private val replay = setOf("m6replay", "w9replay", "6terreplay", "fun_radio", "rtl2")

        private const val URL_TEMPLATE_PODCAST = "http://www.6play.fr/%s-p_%d/"
        private const val PROGRAM_CODE_SELECTOR = "program.programsById.%d.code"
        private const val PROGRAM_ID_SELECTOR = "program.programsById"
        private const val VIDEO_BY_ID_SELECTOR = "video.programVideoById[*]"

        fun isFrom6Play(url: String?) =
                if (url != null && url.contains("www.6play.fr/")) 1
                else Integer.MAX_VALUE
    }
}
